_posts/2016-04-26-aws-and-consul.md

```diff
@@ -0,0 +1,175 @@
+    
+    tag {
+        key = "Name"
+        value = "consul"
+        propagate_at_launch = true
+    }
+}
+
+# Define it's Launch Configuration
+resource "aws_launch_configuration" "consul" {
+    image_id  = "ami-abc123"
+    instance_type = "t2.medium"
+    key_name = "deploy-key"
+    security_groups = ["sg-abc123"]
+
+    # Make sure our instance has a policy which grants read-only access to EC2 API
+    iam_instance_profile = "arn:aws:iam::0123456789:some_profile/ec2"
+
+    # More on user_data later on
+    user_data = "${template_file.consul_userdata.rendered}"
+
+    root_block_device {
+        volume_type = "gp2"
+        volume_size = 10
+    }
+
+    lifecycle {
+        create_before_destroy = true
+    }
+}
+```
+**Nb.** This is simplified a lot, in a real example, much would be variable-ised, but for the sake of a self-contained example I have used real (made-up) values instead of variables.
+
+Nothing too crazy here, we have an ASG + LC which will launch 5 instances based off our AMI. The cleverness here, which is a bit specific to Terraform but can be replicated in other ways using Cloud Formation for example, is the `create_before_destroy = true` statement. So what does that mean? If you change the AMI id, it will create a new ASG+LC alongside the old one. Only when the new instances are attached to the ELB (Elastic Load Balancer) and the ELB declares them healthy will the old instances be torn down. 
+
+### User Data
+If we build this now, we will get a set of 5 Consul nodes, we won't get a Cluster. As I alluded to before, we need to do a little post-launch config to get the nodes to start talking to each other. If we make sure our Consul AMI has the AWS CLI tools installed, its fairly easy to use that to query AWS and ask for all instances tagged 'consul' and get their IP addresses and then shove those values into Consul's config file and finally, start the service.
+
+```
+#!/bin/bash
+
+internalIP=$(curl http://169.254.169.254/latest/meta-data/local-ipv4)
+instanceID=$(curl http://169.254.169.254/latest/meta-data/instance-id)
+hostname="consul-$${instanceID#*-}"
+
+hostnamectl set-hostname $hostname
+
+aws ec2 describe-instances --region eu-west-1 --filters 'Name=tag:Name,Values=consul' 'Name=instance-state-name,Values=running' | jq -r '.Reservations[].Instances[].PrivateIpAddress' > /tmp/instances
+
+while read line;
+do
+ if [ "$line" != "$internalIP" ]; then
+    echo "Adding address $line"
+    cat /etc/consul/000-consul.json | jq ".retry_join += [\"$line\"]" > /tmp/$${line}-consul.json
+
+    if [ -s /tmp/$${line}-consul.json ]; then
+        cp /tmp/$${line}-consul.json /etc/consul/000-consul.json
+    fi
+ fi
+done < /tmp/instances
+
+rm -f /tmp/instances
+
+# Clear any old state from the build process
+rm -rf /var/lib/consul/*
+
+systemctl start consul
+```
+**Nb.**  The double dollar notation here is because Terraform needs ${blah} escaped. If you are not using Terraform, then the double dollar is not needed.
```

Comment:
I would present this script without the Terraform escaping, then mention in the NB that this script won't work in Terraform because it needs the ${blah} escaping - makes the script itself easier to understand when reading.